app/service/PdfService.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package service
+
+import connectors.PdfGeneratorServiceConnector
+import javax.inject.{Inject, Singleton}
+import play.api.http.Status._
+import play.api.mvc.{Result, Results}
+import play.twirl.api.Html
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+@Singleton
+class PdfService @Inject()(connector: PdfGeneratorServiceConnector) {
+
+  def generatePdf(filename: String, htmlContent: Html): Future[Result] = {
+
+    connector.generatePdf(htmlContent).map {
+      response =>
+        response.status match {
+          case OK => Results.Ok(response.bodyAsBytes.toArray)
+            .as("application/pdf")
+            .withHeaders("Content-Disposition" -> s"attachment; filename=$filename")
+          case _ => Results.BadRequest(response.body)
```

Comment:
I think this code can be written better
You return 400 regardless of the upstream status (404, 500, etc)

Reply:
Seems a little strange in my opinion to have a ‘Service’ that is handling responses and requests, shouldn’t that be left to the controller and connector?

Reply:
I agree.  I think it stems from using the WSClient in the connector.  Once I work out how to use the HttpClient I think I can get the connector to return something more useful - e.g. an Array[Byte] with which the service can create the Result.  Then (I'm guessing) I won't have to worry about bad responses - they'll just bubble up.

Reply:
yes, do not handle errors here, do it in the controller

check also the operational UI in case

Reply:
> I agree. I think it stems from using the WSClient in the connector. Once I work out how to use the HttpClient I think I can get the connector to return something more useful - e.g. an Array[Byte] with which the service can create the Result. Then (I'm guessing) I won't have to worry about bad responses - they'll just bubble up.

I think using the http verbs client like the other connectors use should make it easier as you can specify a return type e.g. `client.GET[SomePDFType](url)`